omni-relayer/src/workers/solana.rs

```diff
@@ -0,0 +1,290 @@
+                        }
+                    };
+                }
+            }));
+        }
+
+        join_all(handlers).await;
+
+        tokio::time::sleep(tokio::time::Duration::from_secs(
+            utils::redis::SLEEP_TIME_AFTER_EVENTS_PROCESS_SECS,
+        ))
+        .await;
+    }
+}
+
+#[derive(Debug, serde::Serialize, serde::Deserialize)]
+pub struct InitTransferWithTimestamp {
+    pub amount: u128,
+    pub token: String,
+    pub recipient: String,
+    pub fee: u128,
+    pub native_fee: u64,
+    pub emitter: String,
+    pub sequence: u64,
+    pub creation_timestamp: i64,
+    pub last_update_timestamp: Option<i64>,
+}
+
+pub async fn finalize_transfer(
+    config: config::Config,
+    redis_client: redis::Client,
+    connector: Arc<OmniConnector>,
+) -> Result<()> {
+    let redis_connection = redis_client.get_multiplexed_tokio_connection().await?;
+
+    loop {
+        let mut redis_connection = redis_connection.clone();
+
+        let events = match utils::redis::get_events(
+            &mut redis_connection,
+            utils::redis::SOLANA_INIT_TRANSFER_EVENTS.to_string(),
+        )
+        .await
+        {
+            Some(events) => events,
+            None => {
+                tokio::time::sleep(tokio::time::Duration::from_secs(
+                    utils::redis::SLEEP_TIME_AFTER_EVENTS_PROCESS_SECS,
+                ))
+                .await;
+                continue;
+            }
+        };
+
+        let mut handlers = Vec::new();
+
+        for (key, event) in events {
+            if let Ok(init_transfer_with_timestamp) =
+                serde_json::from_str::<InitTransferWithTimestamp>(&event)
+            {
+                let handler = handle_init_transfer_event(
+                    config.clone(),
+                    connector.clone(),
+                    redis_connection.clone(),
+                    key.clone(),
+                    init_transfer_with_timestamp,
+                );
+                handlers.push(tokio::spawn(handler));
+            }
+        }
+
+        join_all(handlers).await;
+
+        tokio::time::sleep(tokio::time::Duration::from_secs(
+            utils::redis::SLEEP_TIME_AFTER_EVENTS_PROCESS_SECS,
+        ))
+        .await;
+    }
+}
+
+async fn handle_init_transfer_event(
+    config: config::Config,
+    connector: Arc<OmniConnector>,
+    mut redis_connection: redis::aio::MultiplexedConnection,
+    key: String,
+    init_transfer_with_timestamp: InitTransferWithTimestamp,
+) {
+    let current_timestamp = chrono::Utc::now().timestamp();
+
+    if current_timestamp
+        - init_transfer_with_timestamp
+            .last_update_timestamp
+            .unwrap_or_default()
+        < utils::redis::CHECK_INSUFFICIENT_FEE_TRANSFERS_EVERY_SECS
+    {
+        return;
+    }
+
+    if current_timestamp - init_transfer_with_timestamp.creation_timestamp
```

Comment:
Let’s move this check to the end of the function. This will ensure that the function is executed at least once for the event.